Improve job stat history retention job

kpan2034 opened the pull request:
Previously a delete query was run on the `bgw_job_stat_history` table to
remove job history. This could be slow once the job history table became
very large. We now use a temporary table to keep recent history and
truncate the job history table instead. We also introduce a
`job_history_bsearch` function which finds the ids of the jobs that need
to be deleted using binary search.

This PR is based on experiments done
previously (https://github.com/timescale/timescaledb-extras/pull/47).

Disable-check: approval-count

fabriziomello commented on sql/job_stat_history_log_retention.sql:
```diff
@@ -4,18 +4,128 @@
-    DELETE
-    FROM _timescaledb_internal.bgw_job_stat_history
-    WHERE execution_finish < (now() - drop_after);
+  IF NOT FOUND THEN
+    RAISE EXCEPTION 'system job 3 (job execution history retention) not found';
+    RETURN 0;
+  END IF;
```
```suggestion
```
Useless

fabriziomello commented on sql/job_stat_history_log_retention.sql:
```diff
@@ -4,18 +4,128 @@
+  WHERE id >= id_found AND execution_finish >= search_point
+  ORDER BY id;
+
+  ALTER TABLE _timescaledb_internal.bgw_job_stat_history
+    DROP CONSTRAINT IF EXISTS bgw_job_stat_history_pkey;
+
+  DROP INDEX IF EXISTS _timescaledb_internal.bgw_job_stat_history_job_id_idx;
```
```suggestion
```
I think we shouldn't do DDL inside this job cause is easy to forget to change it if for some reason in the future we change something in this table

fabriziomello commented on sql/job_stat_history_log_retention.sql:
```diff
@@ -4,23 +4,97 @@
 -- A retention policy is set up for the table _timescaledb_internal.job_errors (Error Log Retention Policy [2])
 -- By default, it will run once a month and and drop rows older than a month.
+
+CREATE OR REPLACE FUNCTION _timescaledb_functions.job_history_bsearch(search_point TIMESTAMPTZ) RETURNS BIGINT
```
Do we need isolation tests for it since we're locking the table?

kpan2034 replied:
Probably a good idea. Will do this.

fabriziomello commented:
Please write a better PR and commit message describing how we improved things in this PR.

fabriziomello commented on sql/job_stat_history_log_retention.sql:
```diff
@@ -4,23 +4,97 @@
-    FROM _timescaledb_internal.bgw_job_stat_history
-    WHERE execution_finish < (now() - drop_after);
+  -- We need to prevent concurrent changes on this table when running this retention job
+  LOCK TABLE _timescaledb_internal.bgw_job_stat_history IN SHARE ROW EXCLUSIVE MODE;
```
```suggestion
  LOCK TABLE _timescaledb_internal.bgw_job_stat_history IN EXCLUSIVE MODE;
```
I've made a mistake the correct explicit lock here is `EXCLUSIVE MODE`: https://www.postgresql.org/docs/current/explicit-locking.html

fabriziomello replied:
Just wondering if we should fallback to a `DELETE` if for some reason the lock is not taken. We can set a local lock timeout (200ms or 300ms) and if the lock is not taken then we fallback to a `DELETE` instead. We keep use the `job_history_bsearch` because with it we can use the primary key index to filter rows (the root cause of this implementation since the current `DELETE` filter the rows based on the `execution_finish` columns that is not indexed).

kpan2034 replied:
I changed it to an AccessExclusiveLock and added a lock timeout.

kpan2034 commented on sql/job_stat_history_log_retention.sql:
```diff
@@ -4,23 +4,97 @@
 
 -- A retention policy is set up for the table _timescaledb_internal.job_errors (Error Log Retention Policy [2])
 -- By default, it will run once a month and and drop rows older than a month.
+
```
Done.

kpan2034 commented:
I realized we had a constraint on `execution_finish` when copying rows into the temp table, which I removed. This required some minor changes in `job_history_bsearch`.
 
There was also a bug in the `id_middle_next` logic, which I now fixed. I also added test coverage for the `job_history_bsearch` function directly.

fabriziomello commented on tsl/test/expected/bgw_job_stat_history.out:
```diff
@@ -413,3 +413,346 @@ SELECT _timescaledb_functions.stop_background_workers();
+(8, 905, 9005, true, now() - interval '2 weeks', now() - interval '2 weeks' + interval '5 minutes', '{}'),
+(9, 906, 9006, true, now() - interval '1 week', now() - interval '1 week' + interval '5 minutes', '{}'),
+(10, 907, 9007, true, now() - interval '3 days', now() - interval '3 days' + interval '5 minutes', '{}');
+ERROR:  INSERT has more expressions than target columns at character 229
```
Is this error expected?

kpan2034 replied:
Sorry, missed this when adding tests, fixed

pnthao commented on sql/job_stat_history_log_retention.sql:
```diff
@@ -4,23 +4,113 @@
+  id_lower BIGINT;
+  id_upper BIGINT;
+  id_middle BIGINT DEFAULT 0;
+  id_middle_next BIGINT;
```
is_middle_next, range_start and range_end are not used, aren't they?

kpan2034 replied:
Removed.

fabriziomello commented on sql/job_stat_history_log_retention.sql:
```diff
@@ -4,23 +4,97 @@
 -- A retention policy is set up for the table _timescaledb_internal.job_errors (Error Log Retention Policy [2])
 -- By default, it will run once a month and and drop rows older than a month.
+
+CREATE OR REPLACE FUNCTION _timescaledb_functions.job_history_bsearch(search_point TIMESTAMPTZ) RETURNS BIGINT
```
> On another thought, why don't we partition the log table based on time, and just drop the old partitions/chunks?

It was a very long discussion in the past and we didn't reached and agreement so we decided to create a regular table for it. If you want I can share with you a very old design document about it.

pnthao replied:
> > On another thought, why don't we partition the log table based on time, and just drop the old partitions/chunks?
> 
> It was a very long discussion in the past and we didn't reached and agreement so we decided to create a regular table for it. If you want I can share with you a very old design document about it.

Thanks @fabriziomello for the information. Yes I'd love to get some context about this to understand more.

kpan2034 commented:
Updated the PR to use a fixed timestamp for the tests rather than `now()`, to make them deterministic.

fabriziomello commented on sql/job_stat_history_log_retention.sql:
```diff
@@ -4,23 +4,110 @@
+    id_found BIGINT;
 BEGIN
-    drop_after := config->>'drop_after';
+  SET LOCAL lock_timeout = '5s';
```
```suggestion
  PERFORM set_config('lock_timeout', coalesce(config->>'lock_timeout', '5s'), true /* is local */);
```
We can add a test for it in a following PR. The idea is to be able to set a config parameter with `lock_timeout` key. Just wondering if for some reason we can have a problem in our fleet with the fixed 5 seconds default value.

kpan2034 replied:
Updated the PR. Need to add a proper test for it. We should probably also add an isolation test to check the lock timeout works correctly. Can be a follow-up PR.

fabriziomello commented on tsl/test/isolation/specs/bgw_job_stat_history_retention_isolation.spec:
```diff
@@ -0,0 +1,63 @@
+
+step "s2_lock_table" {
+    BEGIN;
+    LOCK TABLE _timescaledb_internal.bgw_job_stat_history IN ACCESS EXCLUSIVE MODE;
```
I think here you can do a weak lock instead to make evident that the retention will wait for any lock on the job history table. An `ACCESS SHARE MODE` is enough.

fabriziomello reviewed (approved):
Left some comments but LGTM.
